### PR TITLE
Use PropertyDeclarationBlock in a DOMRefCell everywhere.

### PR DESCRIPTION
**Pull request by SimonSapin**
**Do not merge** (yet).

r? @emilio 

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13176)

### CODE REVIEW

File: components/style/domrefcell.rs [hunk 1]
```diff
@@ -11,12 +11,19 @@ use thread_state;
...
 /// This extends the API of `core::cell::RefCell` to allow unsafe access in
 /// certain situations, with dynamic checking in debug builds.
-#[derive(Clone)]
+#[derive(Clone, PartialEq, Debug)]
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub struct DOMRefCell<T> {
     value: RefCell<T>,
 }
 
+// FIXME: These two impls make promises that are not quite true,
```
Comment (emilio) [inline components/style/domrefcell.rs]: Can you open an issue about this, and point to it, so there's a central place for discussing this?

File: components/style/keyframes.rs [hunk 2]
```diff
@@ -150,11 +153,13 @@ pub struct KeyframesAnimation {
 ///
 /// In practice, browsers seem to try to do their best job at it, so we might
 /// want to go through all the actual keyframes and deduplicate properties.
+#[allow(unsafe_code)]
 fn get_animated_properties(keyframe: &Keyframe) -> Vec<TransitionProperty> {
```
Comment (emilio) [inline components/style/keyframes.rs]: This is only called from `add_stylesheet`, which is called from `Stylist::update`, so this is safe.

File: components/style/selector_matching.rs [hunk 3]
```diff
@@ -402,8 +404,9 @@ impl Stylist {
         debug!("author normal: {:?}", relations);
 
         // Step 4: Normal style attributes.
-        if let Some(sa)  = style_attribute {
-            if sa.any_normal() {
+        if let Some(sa) = style_attribute {
+            // FIXME: Is this thread-safe?
+            if unsafe { sa.borrow_for_layout() }.any_normal() {
```
Comment (emilio) [inline components/style/selector_matching.rs]: Yes, since this is only called from style resolution code.

**Comment by emilio** [comment]
This is a way to scary PR to land without https://github.com/servo/servo/issues/13127... I think most uses are fine (the one I haven't checked thoroughly is the iterator one).

File: components/style/domrefcell.rs (hunk 1 above)
Comment (SimonSapin) [inline components/style/domrefcell.rs]: https://github.com/servo/servo/issues/13177

**Comment by SimonSapin** [comment]
@pcwalton, could you have a look at the thread-safety of this?

**Comment by SimonSapin** [comment]
The try failure is #11618.

**Comment by SimonSapin** [comment]
**Do not merge** (yet).

I’ve been trying a different design where you pass around a token that says what thread you’re on, and only creating a token is `unsafe` and should be done e.g. near the thread spawn or bindings call. It turns out that `LayoutThread::tick_animations` eventually calls `compute_style_for_animation_step` which calls `cascade` which in this PR uses `borrow_for_layout` on property declaration blocks. But it doesn’t look like the script thread is blocked while that happens, so there could be a data race with CSSOM mutating these declarations.

Based on discussion with @brittles though, it looks like to implement `requestAnimationFrame` correctly we’ll need to synchronize animations with the script thread anyway.

**Pull request closed without merging**